### ISSUE TITLE
Tooltip.js - closeOnClickOutside doesn't work on IOS - both Chrome & Safari

### ISSUE THREAD

**Issue by jwilson114** [FezVrasta/popper.js#783]
closeOnClickOutside doesn't work on IOS as expected on both Chrome & Safari but works great on Android devices.

Meaning, when a user clicks anywhere outside the tooltip container it doesn't close except if the user clicks on the tooltip activator. 

Is there a fix for this? Thanks

**Comment by FezVrasta** [comment]
Thank you for your issue.

Unfortunately I can't answer to all the reported issues without a little help from you.
Please, make sure to follow the **issues template** when you submit a new issue, in this way I will already have all the informations needed to understand the problem and help you with it.

I'm closing this issue for now, but feel free to open a new one making sure to follow all the instructions!

Thank you for your understanding.

**Comment by bohdan-didukh** [comment]
@FezVrasta,

this issue is reproducible for me. I see that tooltip.js using the `mousedown` event, which is not firing on mobile devices. You may try to change it to `click`.

```
// schedule hide tooltip
    oppositeEvents.forEach(event => {
      const func = evt => {
        if (evt.usedByTooltip === true) {
          return;
        }
        this._scheduleHide(reference, options.delay, options, evt);
      };
      this._events.push({ event, func });
      reference.addEventListener(event, func);
      if (event === 'click' && options.closeOnClickOutside) {
        document.addEventListener('mousedown', e => {
          if (!this._isOpening) {
            return;
          }
          const popper = this.popperInstance.popper;
          if (reference.contains(e.target) ||
              popper.contains(e.target)) {
            return;
          }
          func(e);
        }, true);
      }
    });
```

Thank you!

**Comment by davidferguson** [comment]
@bohdan-didukh `click` doesn't work either, for reasons best explained in [this article](https://www.quirksmode.org/blog/archives/2014/02/mouse_event_bub.html). This also explains why `mousedown` isn't working - it's mobile Safari being a bit odd.

I think the solution is therefore to use one of the `touch` events. `touchstart` does *work*, but means that if you just scroll the page on mobile, the popup will disappear, which is not as intended.

Therefore I think the best solution is to capture `touchstart`, `touchmove` and `touchend`, and use those to work out whether this is a drag or a tap. If it's a tap, it should hide the popup, and if not (likely a scroll, zoom, etc.), it should ignore.

I've implemented this as below (this is directly editing popper.js from unpkg), gives the expected behaviour for iOS on my site.

```js
if (event === 'click' && options.closeOnClickOutside) {
  function hideFunc (e) {
    if (!_this2._isOpening) {
      return;
    }
    var popper = _this2.popperInstance.popper;
    if (reference.contains(e.target) || popper.contains(e.target)) {
      return;
    }
    func(e);
  }

  // for mouse devices
  document.addEventListener('mousedown', hideFunc, true);

  // fix for safari not liking mousedown, see #783
  let touchIsTap = false;
  document.addEventListener('touchstart', function (e) {
    touchIsTap = true;
  }, true);
  document.addEventListener('touchmove', function (e) {
    touchIsTap = false;
  }, true);
  document.addEventListener('touchend', function (e) {
    if (touchIsTap) hideFunc(e);
  }, true);
}
```

@FezVrasta I was going to submit a PR for this, but I'm not familiar with jasmine for testing, so I thought I'd leave that up to someone else to do.
